File: RDMPEG/RDMPEGDecoder/RDMPEGFrames/RDMPEGFrames.m
//
//  RDMPEGFrames.m
//  RDMPEG
//
//  Created by Serhii Alpieiev on 10/3/17.
//  Copyright © 2017 Readdle. All rights reserved.
//

#import "RDMPEGFrames.h"



NS_ASSUME_NONNULL_BEGIN

@interface RDMPEGFrame ()

@property (nonatomic, assign) NSTimeInterval position;
@property (nonatomic, assign) NSTimeInterval duration;
@end

@implementation RDMPEGFrame

@end



@interface RDMPEGAudioFrame ()

@property (nonatomic, strong) NSData *samples;

@end

@implementation RDMPEGAudioFrame

- (RDMPEGFrameType)type {
    return RDMPEGFrameTypeAudio;
}

@end



@interface RDMPEGVideoFrame ()

@property (nonatomic, assign) NSUInteger width;
@property (nonatomic, assign) NSUInteger height;

@end

@implementation RDMPEGVideoFrame

- (RDMPEGFrameType)type {
    return RDMPEGFrameTypeVideo;
}

@end



@interface RDMPEGVideoFrameBGRA ()

@property (nonatomic, assign) NSUInteger linesize;
@property (nonatomic, strong) NSData *bgra;

@end

@implementation RDMPEGVideoFrameBGRA

- (nullable UIImage *)asImage {
    CGColorSpaceRef colorSpace = CGColorSpaceCreateDeviceRGB();
    if (colorSpace == NULL) {
        return nil;
    }
    
    CGDataProviderRef provider = CGDataProviderCreateWithCFData((__bridge CFDataRef)(self.bgra));
    
    CGImageRef imageRef = CGImageCreate(self.width,
                                        self.height,
                                        8,
                                        32,
                                        self.linesize,
                                        colorSpace,
                                        kCGBitmapByteOrder32Little | kCGImageAlphaNoneSkipFirst,
                                        provider,
                                        NULL,
                                        YES,
                                        kCGRenderingIntentDefault);
    
    UIImage *image = [UIImage imageWithCGImage:imageRef];
    
    CGImageRelease(imageRef);
    CGDataProviderRelease(provider);
    CGColorSpaceRelease(colorSpace);
    
    return image;
}

@end



@interface RDMPEGVideoFrameYUV ()

@property (nonatomic, strong) NSData *luma;
@property (nonatomic, strong) NSData *chromaB;
@property (nonatomic, strong) NSData *chromaR;

@end

@implementation RDMPEGVideoFrameYUV

@end



@interface RDMPEGArtworkFrame ()

@property (nonatomic, strong) NSData *picture;

@end

@implementation RDMPEGArtworkFrame

- (RDMPEGFrameType)type {
    return RDMPEGFrameTypeArtwork;
}

- (nullable UIImage *)asImage {
    CGDataProviderRef provider = CGDataProviderCreateWithCFData((__bridge CFDataRef)(self.picture));
    
    CGImageRef imageRef = CGImageCreateWithJPEGDataProvider(provider,
                                                            NULL,
                                                            YES,
                                                            kCGRenderingIntentDefault);
    
    UIImage *image = [UIImage imageWithCGImage:imageRef];
    
    CGImageRelease(imageRef);
    CGDataProviderRelease(provider);
    
    return image;
    
}
@end



@interface RDMPEGSubtitleFrame ()

@property (nonatomic, strong) NSString *text;

@end

@implementation RDMPEGSubtitleFrame

- (RDMPEGFrameType)type {
    return RDMPEGFrameTypeSubtitle;
}

@end

NS_ASSUME_NONNULL_END

File: RDMPEG/RDMPEGDecoder/RDMPEGFrames/RDMPEGFrames.swift
//
//  RDMPEGFrames.swift
//  RDMPEG
//
//  Created by Max Berezhnoy on 15/07/2024.
//  Copyright © 2024 Readdle. All rights reserved.
//

// swiftlint:disable file_types_order

import UIKit

@objc
public enum RDMPEGFrameType: UInt {
    case audio
    case video
    case artwork
    case subtitle
}

@objcMembers
public class RDMPEGFrame: NSObject {
    public let type: RDMPEGFrameType
    public var position: TimeInterval
    public var duration: TimeInterval

    public init(type: RDMPEGFrameType, position: TimeInterval, duration: TimeInterval) {
        self.type = type
        self.position = position
        self.duration = duration
        super.init()
    }
}

@objcMembers
public class RDMPEGAudioFrame: RDMPEGFrame {
    public var samples: Data

    public init(position: TimeInterval, duration: TimeInterval, samples: Data) {
        self.samples = samples
        super.init(type: .audio, position: position, duration: duration)
    }
}

@objcMembers
public class RDMPEGVideoFrame: RDMPEGFrame {
    public var width: UInt
    public var height: UInt

    public init(position: TimeInterval, duration: TimeInterval, width: UInt, height: UInt) {
        self.width = width
        self.height = height
        super.init(type: .video, position: position, duration: duration)
    }
}

@objcMembers
public class RDMPEGVideoFrameBGRA: RDMPEGVideoFrame {
    public var linesize: UInt
    public var bgra: Data

    public init(position: TimeInterval, duration: TimeInterval, width: UInt, height: UInt, bgra: Data, linesize: UInt) {
        self.bgra = bgra
        self.linesize = linesize
        super.init(position: position, duration: duration, width: width, height: height)
    }

    public func asImage() -> UIImage? {
        guard let colorSpace = CGColorSpace(name: CGColorSpace.sRGB) else { return nil }

        let provider = CGDataProvider(data: bgra as CFData)

        guard let imageRef = CGImage(
            width: Int(width),
            height: Int(height),
            bitsPerComponent: 8,
            bitsPerPixel: 32,
            bytesPerRow: Int(linesize),
            space: colorSpace,
            bitmapInfo: CGBitmapInfo(
                rawValue: CGImageAlphaInfo.noneSkipFirst.rawValue | CGBitmapInfo.byteOrder32Little.rawValue
            ),
            provider: provider!,
            decode: nil,
            shouldInterpolate: true,
            intent: .defaultIntent
        ) else { return nil }

        return UIImage(cgImage: imageRef)
    }
}

@objcMembers
public class RDMPEGVideoFrameYUV: RDMPEGVideoFrame {
    public var luma: Data
    public var chromaB: Data
    public var chromaR: Data

    public init(
        position: TimeInterval,
        duration: TimeInterval,
        width: UInt,
        height: UInt,
        luma: Data,
        chromaB: Data,
        chromaR: Data
    ) {
        self.luma = luma
        self.chromaB = chromaB
        self.chromaR = chromaR
        super.init(position: position, duration: duration, width: width, height: height)
    }
}

@objcMembers
public class RDMPEGArtworkFrame: RDMPEGFrame {
    public var picture: Data

    public init(picture: Data) {
        self.picture = picture
        super.init(type: .artwork, position: 0, duration: 0)
    }

    public func asImage() -> UIImage? {
        guard let provider = CGDataProvider(data: picture as CFData) else { return nil }

        guard let imageRef = CGImage(jpegDataProviderSource: provider,
                                     decode: nil,
                                     shouldInterpolate: true,
                                     intent: .defaultIntent) else { return nil }

        return UIImage(cgImage: imageRef)
    }
}

@objcMembers
public class RDMPEGSubtitleFrame: RDMPEGFrame {
    public var text: String

    public init(position: TimeInterval, duration: TimeInterval, text: String) {
        self.text = text
        super.init(type: .subtitle, position: position, duration: duration)
    }
}

// swiftlint:enable file_types_order
